console-subscriber/src/visitors.rs

```diff
@@ -0,0 +1,369 @@
+//! fields from tracing metadata and producing the parts
+//! needed to construct `Event` instances.
+
+use super::{AttributeUpdate, AttributeUpdateOp, Readiness, WakeOp};
+use console_api as proto;
+use proto::field::Name as PbFieldName;
+use proto::field::Value as PbFieldValue;
+use proto::resources::resource;
+use tracing_core::{
+    field::{self, Visit},
+    span,
+};
+
+/// Used to extract the fields needed to construct
+/// an Event::Resource from the metadata of a tracing span
+/// that has the following shape:
+///
+/// tracing::trace_span!(
+///     "runtime.resource",
+///     concrete_type = "Sleep",
+///     kind = "timer",
+/// );
+///
+/// Fields:
+/// concrete_type - indicates the concrete rust type for this resource
+/// kind - indicates the type of resource (i.e. timer, sync, io )
+#[derive(Default)]
+pub(crate) struct ResourceVisitor {
+    concrete_type: Option<String>,
+    kind: Option<resource::Kind>,
+}
+
+/// Used to extract all fields from the metadata
+/// of a tracing span
+pub(crate) struct FieldVisitor {
+    fields: Vec<proto::Field>,
+    meta_id: proto::MetaId,
+}
+
+/// Used to extract the fields needed to construct
+/// an Event::AsyncOp from the metadata of a tracing span
+/// that has the following shape:
+///
+/// tracing::trace_span!(
+///     "runtime.resource.async_op",
+///     source = "Sleep::new_timeout",
+/// );
+///
+/// Fields:
+/// source - the method which has created an instance of this async operation
+#[derive(Default)]
+pub(crate) struct AsyncOpVisitor {
+    source: Option<String>,
+}
+
+/// Used to extract the fields needed to construct
+/// an Event::Waker from the metadata of a tracing span
+/// that has the following shape:
+///
+/// tracing::trace!(
+///     target: "tokio::task::waker",
+///     op = "waker.clone",
+///     task.id = id.into_u64(),
+/// );
+///
+/// Fields:
+/// task.id - the id of the task this waker will wake
+/// op - the operation associated with this waker event
+#[derive(Default)]
+pub(crate) struct WakerVisitor {
+    id: Option<span::Id>,
+    op: Option<WakeOp>,
+}
+
+/// Used to extract the fields needed to construct
+/// an Event::PollOp from the metadata of a tracing event
+/// that has the following shape:
+///
+/// tracing::trace!(
+///     target: "tokio::resource::poll_op",
+///     op_name = "poll_elapsed",
+///     readiness = "pending"
+/// );
+///
+/// Fields:
+/// op_name - the name of this resource poll operation
+/// readiness - the result of invoking this poll op, describing its readiness
+#[derive(Default)]
+pub(crate) struct PollOpVisitor {
+    op_name: Option<String>,
+    readiness: Option<Readiness>,
+}
+
+/// Used to extract the fields needed to construct
+/// an Event::StateUpdate from the metadata of a tracing event
+/// that has the following shape:
+///
+/// tracing::trace!(
+///     target: "tokio::resource::state_update",
+///     duration = "attribute_name",
+///     value = 10,
+///     unit = "ms",
+///     op = "ovr",
+/// );
```

Comment:
The way that state updates are recorded currently feels a little overly complex to me. We are looking for four separate fields named `attribute_name`, `value`, `op`, and `unit`, but `tracing` already has a mechanism for representing key-value pairs. Why not simply look for fields named `unit` and op`, and then treat any remaining field as the value, and use its name as the name of the attribute?

We could potentially even do something a bit fancier to allow a single event to update multiple resource states, if we looked for clusters of fields named `<NAME>`,  `<NAME>.unit`, and `<NAME>.op`? That way, we could record one event like
```rust
tracing::trace!(
    target: "runtime::resource::state_update", 
    read = 256, read.unit = "bytes", read.op = "add",
    to_write = 1024, to_write.unit = "bytes", to_write.op = "sub",
);
```
or similar?

This might be slightly overengineered, but it seems like it would be nice to reuse the existing mechanism in `tracing`, and in the wire proto, for representing field names. In particular, if we do it this way, we can just use the existing proto format for representing field names as metadata indices, instead of having to add a new concept of "attributes", which are like tracing fields but also not the same...